pkg/distributor/receiver/shim.go

```diff
@@ -0,0 +1,318 @@
+
+	"github.com/grafana/loki/pkg/logproto"
+	util_log "github.com/grafana/loki/pkg/util/log"
+)
+
+const (
+	LogFormatJSON   = "json"
+	LogFormatLogfmt = "logfmt"
+)
+
+var (
+	metricPushDuration = promauto.NewHistogram(prom_client.HistogramOpts{
+		Namespace: "loki",
+		Name:      "distributor_push_duration_seconds",
+		Help:      "Records the amount of time to push a batch to the ingester.",
+		Buckets:   prom_client.DefBuckets,
+	})
+)
+
+type BatchPusher interface {
+	Push(ctx context.Context, req *logproto.PushRequest) (*logproto.PushResponse, error)
+}
+
+type receiversShim struct {
+	services.Service
+
+	receivers    []component.Receiver
+	pusher       BatchPusher
+	logger       log.Logger
+	format       string
+	drainTimeout time.Duration
+}
+
+func (r *receiversShim) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+// New
+// handler "/v1/logs" go.opentelemetry.io/collector/receiver/otlpreceiver/otlp.go:229
+func New(receiverCfg map[string]interface{}, pusher BatchPusher, receiverFormat string, receiverDrainTimeout time.Duration, middleware Middleware, logLevel logging.Level) (services.Service, error) {
+	shim := &receiversShim{
+		pusher:       pusher,
+		format:       receiverFormat,
+		logger:       util_log.Logger,
+		drainTimeout: receiverDrainTimeout,
+	}
+
+	receiverFactories, err := component.MakeReceiverFactoryMap(
+		otlpreceiver.NewFactory(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	p := config.NewMapFromStringMap(map[string]interface{}{
+		"receivers": receiverCfg,
+	})
+	cfgs, err := configunmarshaler.NewDefault().Unmarshal(p, component.Factories{
+		Receivers: receiverFactories,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+	params := component.ReceiverCreateSettings{TelemetrySettings: component.TelemetrySettings{
+		Logger:         zap.NewNop(),
+		TracerProvider: trace.NewNoopTracerProvider(),
+		MeterProvider:  metric.NewNoopMeterProvider(),
+	}}
+
+	for componentID, cfg := range cfgs.Receivers {
+		factoryBase := receiverFactories[componentID.Type()]
+		if factoryBase == nil {
+			return nil, fmt.Errorf("receiver factory not found for type: %s", componentID.Type())
+		}
+
+		receiver, err := factoryBase.CreateLogsReceiver(ctx, params, cfg, middleware.Wrap(shim))
+		if err != nil {
+			return nil, err
+		}
+
+		shim.receivers = append(shim.receivers, receiver)
+	}
+
+	shim.Service = services.NewIdleService(shim.starting, shim.stopping)
+
+	return shim, nil
+}
+func (r *receiversShim) starting(ctx context.Context) error {
+	for _, receiver := range r.receivers {
+		r.logger.Log("msg", "receiver start")
+		err := receiver.Start(ctx, r)
+		if err != nil {
+			return fmt.Errorf("error starting receiver %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Called after distributor is asked to stop via StopAsync.
+func (r *receiversShim) stopping(_ error) error {
+	// when shutdown is called on the receiver it immediately shuts down its connection
```

Comment:
Is this a particularity of Loki? On the otelcol's side, in-flight requests should have a chance to be completed.

Reply:
related tempo issue：https://github.com/grafana/tempo/issues/530 《Distributors should drain before terminating》
related tempo PR：https://github.com/grafana/tempo/pull/604

for k8s